### PR TITLE
MSC2265: Proposal for mandating case folding when processing e-mail address localparts

### PR DESCRIPTION
**Pull request by babolivier**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/babolivier/msc_email_case/proposals/2265-email-lowercase.md)

### CODE REVIEW

File: proposals/2265-email-lowercase.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+out of their account because of that). So far we've seen that confusion occur
+and lead to troubles of various degrees over several deployments of Synapse and
+Sydent.
+
+## Proposal
+
+This proposal suggests changing the specification of the e-mail 3PID type in
+[the Matrix spec appendices](https://matrix.org/docs/spec/appendices#pid-types)
+to mandate that any e-mail address must be entirely converted to lowercase
+before any processing, instead of only its domain.
```
Comment (KitsuneRal) [inline proposals/2265-email-lowercase.md]: I wonder how much of complication is to mandate lower-case processing (such as lookup and hashing) but case-preserve storing addresses.

Comment (turt2live) [reply proposals/2265-email-lowercase.md]: we ultimately can't tell implementations how to store their data (if they want to spend the extra time converting things to uppercase they can), but the requirement for lookups being lowercase is a fairly strong argument imo

File: proposals/2265-email-lowercase.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+doesn't mandate anything about processing e-mail addresses, other than the fact
+that the domain part must be converted to lowercase, as domain names are case
+insensitive.
+
+On the other hand, most major e-mail providers nowadays process the localparts
+of e-mail addresses as case insensitive. Therefore, most users expect localparts
+to be treated case insensitively, and get confused when it's not. Some users,
+for example, get confused over the fact that registering a 3PID association for
+`john.doe@example.com` doesn't mean that the association is valid for
+`John.Doe@example.com`, and don't expect to be expected to remember the exact
```
Comment (anoadragon453) [inline proposals/2265-email-lowercase.md]:
```suggestion
`John.Doe@example.com`, and don't expect to have to remember the exact
```

File: proposals/2265-email-lowercase.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+This proposal suggests changing the specification of the e-mail 3PID type in
+[the Matrix spec appendices](https://matrix.org/docs/spec/appendices#pid-types)
+to mandate that any e-mail address must be entirely converted to lowercase
+before any processing, instead of only its domain.
+
+## Other considered solutions
+
+A first look at this issue concluded that there was no need to add such a
+mention to the spec, and that it can be considered as an implementation detail.
```
Comment (anoadragon453) [inline proposals/2265-email-lowercase.md]:
```suggestion
mention to the spec, and that it can be considered an implementation detail.
```

File: proposals/2265-email-lowercase.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+However, [MSC2134](https://github.com/matrix-org/matrix-doc/pull/2134) changes
+this: because hashing functions are case sensitive, we need both clients and
+identity servers to follow the same policy regarding case sensitivity.
```
Comment (anoadragon453) [inline proposals/2265-email-lowercase.md]: Additionally, we're looking to make everything hash-based in the future, which will make this proposal even more important.

File: proposals/2265-email-lowercase.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+## Tradeoffs
+
+Implementing this MSC in identity servers would require the databases of
+existing identity servers to be updated in a large part to convert the email
+addresses of existing associations to lowercase, in order to avoid conflicts.
+However, most of this update can usually be done by a single database query (or
+a background job running at startup), so the UX improvement outweights this
```
Comment (anoadragon453) [inline proposals/2265-email-lowercase.md]:
```suggestion
a background job running at startup), so the UX improvement outweighs this
```

File: proposals/2265-email-lowercase.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+trouble.
+
+## Potential issues
+
+Some users might already have two different accounts associated with the same
+e-mail address but with different cases. This appears to happen in a small
+number of cases, however, and can be dealt by the identity server's maintainer.
```
Comment (anoadragon453) [inline proposals/2265-email-lowercase.md]:
```suggestion
number of cases, however, and can be dealt with by the identity server's maintainer.
```

File: proposals/2265-email-lowercase.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+## Tradeoffs
+
+Implementing this MSC in identity servers would require the databases of
+existing identity servers to be updated in a large part to convert the email
+addresses of existing associations to lowercase, in order to avoid conflicts.
+However, most of this update can usually be done by a single database query (or
+a background job running at startup), so the UX improvement outweights this
+trouble.
+
+## Potential issues
```
Comment (anoadragon453) [inline proposals/2265-email-lowercase.md]: What about the case where two email addresses `john.doe@example.com` and `John.Doe@example.com` are registered to two separate user accounts on a homeserver? Is this is problem in the current homeserver landscape or can it just be a "tough luck and ignore"?

Comment (babolivier) [reply proposals/2265-email-lowercase.md]: Oooh good catch. I think it's going to be as much as an issue for homeservers as for identity servers, and remediation solutions would look the same... I'll spell it out in the MSC.

Comment (babolivier) [reply proposals/2265-email-lowercase.md]: Done in https://github.com/matrix-org/matrix-doc/pull/2265/commits/520c76a1cb8e8669ac624be3c36cc606aeb4cba6

Comment (anoadragon453) [reply proposals/2265-email-lowercase.md]: Sure

**Comment by anoadragon453** [comment]
Seems like general consensus overall.

@mscbot fcp merge

**Comment by babolivier** [comment]
> Relatedly: should we consider unicode normalisation, so that (for example) 'ê' (U+00EA, e with circumflex) is treated the same as 'ê' (U+0065 U+0302, e followed by circumflex combining character)?

I guess it depends on whether common email providers treat both characters as the same. I'll do some investigation around that.

> Neither of the above solve the 'French problem' where (traditionally) accents are omitted on upper-case characters, so 'COTE' should be equivalent to 'côté'...

Should it, though, keeping in mind we're only looking at email addresses here? I just checked on both Gmail and Hotmail and neither of them consider `bréndan.abolivier@...` as being the same as `brendan.abolivier@...`, and I'm not aware of any that does.

Otherwise, yes, casefold is probably the way to go, I'll update the proposal to reflect that.

**Comment by richvdh** [comment]
tbh I wasn't aware that gmail let you use non-ascii localparts at all. Certainly being guided by the behaviour of common providers seems like a sensible idea. (also: sorry for not starting a thread.)

File: proposals/2265-email-lowercase.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+`john.doe@example.com` doesn't mean that the association is valid for
+`John.Doe@example.com`, and don't expect to have to remember the exact
+case they used to initially register the association (and sometimes get locked
+out of their account because of that). So far we've seen that confusion occur
+and lead to troubles of various degrees over several deployments of Synapse and
+Sydent.
+
+## Proposal
+
+This proposal suggests changing the specification of the e-mail 3PID type in
```
Comment (richvdh) [inline proposals/2265-email-lowercase.md]: what should be the behaviour of (a) an ID server when I try to /bind an email address that contains un-casefolded characters (b) a HS when I call an endpoint that proxies such a request?

**Comment by babolivier** [comment]
> tbh I wasn't aware that gmail let you use non-ascii localparts at all

Neither was I. For full context, I've tried Thunderbird (+OVH), Roundcube (+OVH), Hotmail and Gmail and only this last one accepted a non-ascii localpart in the recipient's address.

File: proposals/2265-email-lowercase.md (hunk 8 above)
Comment (babolivier) [reply proposals/2265-email-lowercase.md]: Always lowercase/casefold, which is what I meant by "before any processing". I'll try to reword this expression to make it less vague.

File: proposals/2265-email-lowercase.md [hunk 9]
```diff
@@ -33,17 +38,24 @@ However, [MSC2134](https://github.com/matrix-org/matrix-doc/pull/2134) changes
 this: because hashing functions are case sensitive, we need both clients and
 identity servers to follow the same policy regarding case sensitivity.
 
+An initial version of this proposal proposed to mandate lowercasing e-mail
+addresses instead of case folding them, however it was pointed out that this
+solution might not be the best and most future-proof one.
+
+Unicode normalisation was also looked at but judged unnecessary.
```
Comment (richvdh) [inline proposals/2265-email-lowercase.md]: was it? did you happen to check if gmail lets you register `brêndan` and `brêndan` as different addresses?

Comment (babolivier) [reply proposals/2265-email-lowercase.md]: It doesn't let you register either of those:
image

Comment (richvdh) [reply proposals/2265-email-lowercase.md]: ok so I didn't mean "register".

For whatever test you did to determine if it considers `bréndan.abolivier` the same as `brendan.abolivier`: does it consider `brêndan` the same as `brêndan` ?

Comment (babolivier) [reply proposals/2265-email-lowercase.md]: ftr we had a chat about this afterwards on another channel and concluded that it was probably too much of a faff to test for something that's probably too much of an edge case.

File: proposals/2265-email-lowercase.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+### Storing and querying
+
+Most database engines don't support case folding, therefore querying all
+e-mail addresses matching a case folded e-mail address might not be trivial,
+e.g. an identity server querying all associations for `strauss@example.com` when
+processing a `/lookup` request would be expected to also get associations for
+`Strauß@Example.com`.
+
+To address this issue, implementation maintainers are strongly encouraged to
+make e-mail addresses go through a full case folding before storing them.
```
Comment (KitsuneRal) [inline proposals/2265-email-lowercase.md]: Alternatively, store an index-friendly value next to the user-entered one in the database.

File: proposals/2265-email-lowercase.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+case they used to initially register the association (and sometimes get locked
+out of their account because of that). So far we've seen that confusion occur
+and lead to troubles of various degrees over several deployments of Synapse and
+Sydent.
+
+## Proposal
+
+This proposal suggests changing the specification of the e-mail 3PID type in
+[the Matrix spec appendices](https://matrix.org/docs/spec/appendices#pid-types)
+to mandate that, before any processing, e-mail address localparts must go
```
Comment (turt2live) [inline proposals/2265-email-lowercase.md]: why just the localpart? Pretty sure if I email JOHNDOE@GMAIL.COM it works fine.

Comment (babolivier) [reply proposals/2265-email-lowercase.md]: The spec already says the domain part should be treated without case sensitivity. Though it's probably better to extend the casefolding to the whole email address to avoid any confusion.

Comment (babolivier) [reply proposals/2265-email-lowercase.md]: Fixed in 2e2f1c11

File: proposals/2265-email-lowercase.md (hunk 10 above)
Comment (mvgorcum) [reply proposals/2265-email-lowercase.md]: That works for the v1 lookup, but the new lookup method uses hashes, and in future it's possible that only the hashes will remain saved in the database.
As such a full case folding seems like a required step from my perspective.

**Comment by turt2live** [comment]
Merged 🎉
